core/templates/pages/exploration-editor-page/translation-tab/translator-overview/translator-overview.component.spec.ts

```diff
@@ -88,6 +88,10 @@ describe('Translator Overview component', () => {
       ],
       schemas: [NO_ERRORS_SCHEMA],
     }).compileComponents();
+
+    UserExplorationPermissionsService.permissionsPromise = Promise.resolve({
```

Comment:
Hi @Vir-8 -- I don't think we just overwrite existing methods in setup. This could carry over to other tests as well and cause other errors in the future. Typically in such cases we would use a spy/mock.

Reply:
Oh apologies, wasn't aware of the issues with overwriting methods.

For a clear explanation, this suite is not supposed to be testing the permissions service and it seems to be either an internal dependency, or possibly other suites affecting the behaviour - which is why I assume it was being spied on earlier.

I considered that and excluded every suite using the permissions service, the tests still flaked, so I believe it isn't originating from another suite.

Considering it an internal dependency, it is unclear what calls the user exploration permissions service itself as it appears to be deeply nested, and the problem here is that spies on the permissions service as well as the related context service call seem to have absolutely no effect - likely as a result of internal promises from dependencies.

This leads me to the conclusion that mocking the service would be the best option, I hope this clarifies things, what do you think?

Reply:
No, this is still too handwavy. What specifically is causing the error in the stacktrace -- what action flow / lines of code are being called?

Reply:
I thought I already mentioned that above, [`getPermissionsAsync()`](https://github.com/oppia/oppia/blob/9a4370d9126b358b39ed1a155da807a6cd4c0fbb/core/templates/pages/exploration-editor-page/services/user-exploration-permissions.service.ts#L41) in UserExplorationPermissionsService is being called, which calls `fetchPermissionsAsync` of the same service. 

This fetch permissions function calls the [`getPermissionsAsync`](https://github.com/oppia/oppia/blob/9a4370d9126b358b39ed1a155da807a6cd4c0fbb/core/templates/domain/exploration/exploration-permissions-backend-api.service.ts#L40) of ExplorationPermissionsBackendApi service, which includes calling the ContextService and flakes.

It is supposed to resolve at the getPermissionsAsync() call of the first service itself and not continue propagating, but it doesn't currently. Does that help?

Reply:
Ah, I see, yes, it does, thanks. That last explanation is clearer because it focuses directly on the call stack.

So then I have a question. Why do you say it is supposed to resolve at the getPermissionsAsync() call? If the getPermissionsAsync() call has never been called before, then it should go on to populate permissionsPromise -- and then that gets cached for later use. So isn't continuing on to fetchPermissionsAsync() the right thing to do?

Reply:
Because we are unable to track where it's being called from, and as a result causing spying on it to having no effect - I believe it's not possible to continue on to fetchPermissionsAsync(), and out of the scope for this component to do that. Also this component isn't supposed to deal with the permissions at all. Does that make sense?

Reply:
@Vir-8 No, not really. In order to determine whether this is in scope for the test, one has to look at the context for the test. The test happens in the exploration editor page's translation tab, where it is quite reasonable for the code to expect that it is in a place where the context service recognizes it as an exploration.

I don't think it's ok to just say "this is out of scope for this test because I can't figure out what's calling it". When you interact with a page/component on the site, that component might make calls all the way to the backend layer. That's fine and expected. It doesn't mean those calls are out of scope (though for frontend tests you'd probably need to mock the backend responses because there isn't a backend running in that env).

Also, in general, we could spy/mock, but in general it's nicer to avoid that if possible unless we're specifically setting initial data for the test. And usually we do the spy/mock only for simple properties and in very targeted places. In this case it seems to me like we need to just properly initialize whatever is needed for the context service call to return the right value.

Does that make sense?

Reply:
You're right, I shouldn't have jumped to conclusions. I manually verified thoroughly and the flake originates from the [constructor of routerService](https://github.com/oppia/oppia/blob/9a4370d9126b358b39ed1a155da807a6cd4c0fbb/core/templates/pages/exploration-editor-page/services/router.service.ts#L70) which calls the `_changeTab()` function.

This function call proceeds further to call a function of [explorationImprovements service](https://github.com/oppia/oppia/blob/9a4370d9126b358b39ed1a155da807a6cd4c0fbb/core/templates/services/exploration-improvements.service.ts#L115), which then proceeds to follow the call stack I sent above.

I feel that a mock would still be appropriate in this scenario where it's so indirect and not possible to fetch the appropriate permissions otherwise, does that sound OK?

Reply:
@seanlip Well it's because almost all of them either mock the context service, the exploration improvements service, the permissions service or the router service itself.

I found a few components that don't, but they don't seem to do anything special either. Apologies but I am quite frankly not sure what you mean by properly initializing it. Please let me know if I'm missing out on something.

Reply:
Ah, sorry, missed replying to this. I meant the last sentence of my penultimate comment here: https://github.com/oppia/oppia/pull/19998#discussion_r1536859033

To put it another way: in the exploration editor page in the dev server, the context service generally does realize that it's there, and returns the right value when called. What is the context service using in order to figure that out? If it's e.g. reading from the URL, then can we mock the URL? Or can we mock the specific function in contextService that needs to return the value, if the issue is that the service is initialized before you can set the URL, so that we reduce the "surface area" of what is being mocked as much as possible?